File: src/SWA.Ariadne.Logic/IMazeSolver.cs
using System;
using System.Collections.Generic;
using System.Text;
using SWA.Ariadne.Model;

namespace SWA.Ariadne.Logic
{
    #region Delegate definitions

    #endregion

    /// <summary>
    /// The public methods a MazeSolver must offer.
    /// </summary>
    public interface IMazeSolver
    {
        /// <summary>
        /// Reset to the initial state (before the maze is solved).
        /// </summary>
        void Reset();

        /// <summary>
        /// Travel from one visited square to a neighbor square (through an open wall).
        /// </summary>
        /// <param name="sq1">first (previously visited) square</param>
        /// <param name="sq2">next (neighbor) square</param>
        /// <param name="forward">true if the neighbor square was not visited previously</param>
        void Step(out MazeSquare sq1, out MazeSquare sq2, out bool forward);

        /// <summary>
        /// Find a path in the maze from the start to the end point.
        /// </summary>
        void Solve();

        /// <summary>
        /// Write state information to the given StringBuilder.
        /// </summary>
        /// <param name="message"></param>
        void FillStatusMessage(StringBuilder message);
    }
}

File: src.JAVA/src/swa/ariadne/logic/IMazeSolver.java
package swa.ariadne.logic;

import swa.ariadne.model.MazeSquare;

/**
 * The public methods any MazeSolver must offer.
 * 
 * @author Stephan.Wacker@web.de
 */
public interface IMazeSolver
{
    //--------------------- Types

    /**
     * Output parameters of the Step() method.
     */
    public static class StepResult
    {
        /** First (previously visited) square of the step. */
        public MazeSquare sq1;

        /** Next (neighbor) square. */
        public MazeSquare sq2;

        /** True if the neighbor square was not visited previously. */
        public boolean forward;
    }

    //--------------------- Methods

    /**
     * Reset to the initial state (before the maze is solved).
     * Subclasses should call their super class' method first.
     */
    void reset();

    /**
     * Travel from one visited square to a neighbor square (through an open wall).
     * @param out The output parameters: sq1, sq2, forward.
     */
    void step(StepResult out);

    /**
     * Find a path in the maze from the start to the end point.
     */
    void solve();
}
